AAInstrumentation/maximal_relaxation.py

```diff
@@ -0,0 +1,311 @@
+        )
+        self.assemble_file(
+            self.instr_dir.joinpath(
+                file_name.parent, str(0) + str(file_name.stem) + ".bc"
+            ),
+            self.instr_dir.joinpath(
+                file_name.parent, str(0) + str(file_name.stem) + ".o"
+            ),
+        )
+
+        return self.get_hash(
+            self.instr_dir.joinpath(
+                file_name.parent, str(0) + str(file_name.stem) + ".o"
+            )
+        )
+
+    def relax_one_index(
+        self,
+        prefix: list[int],
+        file_name: Path,
+        lower_bound: int,
+        i: int,
+    ):
+        """Runs the instrumentation with the queries in prefix relaxed as
+        well as the queries from lower_bound to i."""
+        self.run_step_single_func(
+            file_name,
+            i,
+            prefix + list(range(lower_bound, i)),
+            self.instrument_aa_recursively,
+        )
+
+        self.assemble_file(
+            self.instr_dir.joinpath(
+                file_name.parent, str(i) + str(file_name.stem) + ".bc"
+            ),
+            self.instr_dir.joinpath(
+                file_name.parent, str(i) + str(file_name.stem) + ".o"
+            ),
+        )
+
+    def maximal_relaxation(
+        self,
+    ):
+        self.generate_baseline()
+
+        files = self.get_baseline_files()
+
+        # compile baseline
+        with Pool(self.proc_count) as p:
+            p.map(self.compile_baseline_file, files)
+
+        required_empty_paths = [
+            self.instr_dir,
+            Path("res/"),
+            Path("alias_queries/"),
+            Path("final_res/"),
+            Path("composed_files/"),
+        ]
+
+        count_per_file = self.get_candidates_per_file(
+            files, self.instrument_aa_recursively
+        )
+
+        for file_name in files:
+            print("==== Next file: " + str(file_name))
+            self.setup_directories(required_empty_paths)
+
+            self.maximal_relaxation_single_file(count_per_file[file_name], file_name)
+
+    def maximal_relaxation_single_file(
+        self,
+        candidate_count,
+        file_name: Path,
+    ):
+        """Returns the maximum number of queries that can be relaxed without
+        impact on the result.
+
+        Args:
+            candidate_count: The number of candidates to consider.
+            file_name: The name of the file to run the algorithm on.
+        """
+        os.makedirs(str((self.instr_dir / file_name).parent), exist_ok=True)
+        print("Number of candidates: " + str(candidate_count))
+
+        self.original_hash = self.get_groundtruth_hash(file_name)
+
+        curr_prefix = []
+        lower_bound = 0
+        upper_bound = candidate_count
+
+        while True:
+            prefix = self.relax(
+                file_name,
+                curr_prefix,
+                lower_bound,
+                upper_bound,
+            )
+
+            new_count = self.get_candidate_count(
```

Comment:
Are you compiling twice here just to get the number?

Reply:
Yes, however, this only executes very rarely. Currently, it is executed exactly once for each file. I think this makes the code easier to read/reason about (i.e. on  method does the instrumentation, one tracks the number of candidates), so I don't think this is a huge issue. What do you think?